### PR TITLE
fix(Core): fix chests not despawning when opened in group (2)

### PR DESCRIPTION
**Pull request by r-o-b-o-t-o**
**This is a fixed version of https://github.com/azerothcore/azerothcore-wotlk/pull/5371, which caused crashes. This PR aims at fixing the same issue, without the crashes.**

## Changes Proposed:
-  Deactivate chests when their last item has been rolled by a group

## Issues Addressed:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/5362
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/5128

Video displaying the current (bugged) behaviour:

https://user-images.githubusercontent.com/24357633/115364584-d211ab00-a1c3-11eb-936a-499f5ec0bc12.mp4

## Tests Performed:

- Tested in-game on Windows 10

## How to Test the Changes:

- Log in on two separate accounts
- With one account, invite the character from the other account in a group
- Go to a chest with both characters (example: `.go o 30950`)
- Make sure that the loot rule is not Free For All
- Open the chest, take all gray/white items, close the chest
- With both characters, roll on the green/blue items
- Verify that the chest is despawning when all items have been rolled for

See example video:

https://user-images.githubusercontent.com/24357633/115364603-d6d65f00-a1c3-11eb-9dd6-565126c5ab38.mp4

## Known Issues and TODO List:
**NOTE**: this is my first C++ PR here, so please review this carefully 😁 if you think it could be fixed in a better way, write a comment and I'll edit the changes.

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/game/Groups/Group.cpp [hunk 1]
```diff
@@ -1494,6 +1494,19 @@ void Group::CountTheRoll(Rolls::iterator rollI, Map* allowedMap)
             item->is_blocked = false;
     }
 
+    {
+        Loot* loot = roll->getLoot();
+        if (loot->isLooted() && loot->sourceGameObject != nullptr)
+        {
+            const GameObjectTemplate* goInfo = loot->sourceGameObject->GetGOInfo();
+            if (goInfo != nullptr && goInfo->type == GAMEOBJECT_TYPE_CHEST)
```
Comment (Kitzunu) [inline src/server/game/Groups/Group.cpp:1502]: I believe you can do this as (goInfo && goInfo->....)

That should check for null without you needing to deny null

**Comment by r-o-b-o-t-o** [comment]
So I'm pretty sure that the crashes originated from the `sourceGameObject` that was previously not initialized with `nullptr`
image
but I've also added a null check on `GetGOInfo()`, just to be sure

File: src/server/game/Groups/Group.cpp (hunk 1 above)
Comment (r-o-b-o-t-o) [reply src/server/game/Groups/Group.cpp:1502]: Yes they're the same thing, I just like being explicit, I believe it makes code more readable

Comment (Kitzunu) [reply src/server/game/Groups/Group.cpp:1502]: I would prefer the other approach as it is something we already use a lot, to keep the same code standard

**Comment by Footman** [comment]
> So I'm pretty sure that the crashes originated from the `sourceGameObject` that was previously not initialized with `nullptr`
> image
> but I've also added a null check on `GetGOInfo()`, just to be sure

Hey yes it was because of intialization. sourceGameObject was not null but still invalid because of this.

File: src/server/game/Groups/Group.cpp [hunk 2]
```diff
@@ -1494,6 +1494,19 @@ void Group::CountTheRoll(Rolls::iterator rollI, Map* allowedMap)
             item->is_blocked = false;
     }
 
+    {
+        Loot* loot = roll->getLoot();
+        if (loot->isLooted() && loot->sourceGameObject)
```
Comment (FrancescoBorzi) [inline src/server/game/Groups/Group.cpp:1499]:
```suggestion
        if (loot && loot->isLooted() && loot->sourceGameObject)
```

Comment (FrancescoBorzi) [inline src/server/game/Groups/Group.cpp:1499]:
```suggestion
        if (auto loot = roll->getLoot() && loot->isLooted() && loot->sourceGameObject)
```

**Comment by pklloveyou** [comment]
recovery https://github.com/azerothcore/azerothcore-wotlk/pull/4413 This will fix the error

**Comment by r-o-b-o-t-o** [comment]
> recovery #4413 This will fix the error

No, reverting this PR does not fix chests not despawning when you need to roll on the last items inside the chest. I had already tried reverting it since you mentioned it on one of the issues for this, and it doesn't change anything.

**Comment by r-o-b-o-t-o** [comment]
Hi @FrancescoBorzi @Kitzunu sorry for the ping guys, but I fixed the issues you mentioned in your reviews, could you update your reviews please? 😄

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
approved! is this ready to go or requires more testing? @r-o-b-o-t-o

**Comment by r-o-b-o-t-o** [comment]
> is this ready to go or requires more testing?

I think it's ready, but I'm pretty sure that I'm the only one that tested the crash fix, so maybe someone else should take a look at it. I already asked @Sedatyf to test it so we should be good soon!

**Comment by ErnestoGonzalezC** [comment]
Tested, it works perfectly. Nice job

**Comment by Sedatyf** [comment]
Thank you @ErnestoGonzalezC! I didn't find time to test it before. Did you encounter any crash or weird issue?

**Comment by ErnestoGonzalezC** [comment]
> Thank you @ErnestoGonzalezC! I didn't find time to test it before. Did you encounter any crash or weird issue?

The truth is that no, in my opinion everything is perfect.

**Comment by FrancescoBorzi** [comment]
thanks all!

**Comment by Kitzunu** [comment]
Thanks =)

**Pull request merged**